### PR TITLE
frontend: Details: Make multi cluster aware

### PR DESCRIPTION
**Pull request by illume**
This makes Details views work with multiple cluster. 

The ?cluster is passed via a URL query parameter. Because in the normal cluster part of the URL is the group of clusters, and the details view needs to know which of the clusters the resource is in.

Here you can see details views being clicked on and the work (even though multiple clusters are active).

https://github.com/user-attachments/assets/3a840c4d-7e73-4dd4-8ef9-4b81d171cb2f

- **frontend: DetailsGrid: Fix for using a specific cluster**
- **frontend: router: Fix createRouteURL to take query params**
- **frontend: KubeObject: Add cluster query param to getDetailsLink**
- **frontend: k8s: Add useClusterFromURLVar to get cluster**
- **frontend: Details: Add support for cluster via URL var**

### CODE REVIEW

**Comment by joaquimrocha** [comment]
Please look into this branch, since I think I had also added a way to llink to resources details without having to create a new URL query param:
https://github.com/headlamp-k8s/headlamp/commits/aggregated-view/

**Review by joaquimrocha — CHANGES_REQUESTED** [review]
I think (haven't double checked) that in https://github.com/headlamp-k8s/headlamp/tree/aggregated-view-original-rebased , I had computed the route directly from the resource in ResourceLink, thus using the known URLs without a cluster param.

**Comment by illume** [comment]
>frontend: Automatically add the cluster param to ResourceLink 0e2f0f3
> IIUC, this replaces that 1st PR from the list you shared earlier
>> That's included already. It adds to the cluster param AFAIK. Which is /c/<cluster>/
>> Thanks for the tip though. Let me try running that branch and see if I can find it that way.
>
> Yes, and that's okay because when you go into a resource details view, you want to see only one cluster.
> When you press back, it goes back to whatever the URL was, including the c/SOMEGROUP

The problem with only relying on this is that:
- the URL has only one cluster in it, which means sharing/using links loses the cluster group
- if a person clicks on another link in the detail view then they lose the cluster group

This PR makes these things work.

**Comment by joaquimrocha** [comment]
I see, and it's a good idea to keep the group context. But for that, I think it makes more sense to do it the other way around: keep the URL for the details view matching the normal details URL (i.e. /c/SINGLE_CLUSTER/resource-kind/my-resource) and use the query param for a group association: `/c/SINGLE_CLUSTER/resource-kind/my-resource?cluster-group=MyGroup`. This way, we accomplish what you want (keep a reference of the group) we don't have 2 different URLs for the same view, and we can eventually lose the query param and the view still works.
Does this make sense?

**Comment by illume** [comment]
All other links would then need to special case where /c/<cluster> means something different?

I don't think your proposal makes sense. Because why would we change the URL to '/c/<clusterGroup>' to just a single cluster just for this?

File: frontend/src/lib/k8s/index.ts [hunk 1]
```diff
@@ -140,6 +140,17 @@ export function useClusterGroup(): string[] {
   return clusterGroup;
 }
 
+/**
+ * Use the cluster name from the URL query parameters if it's there.
+ *
+ * @returns the cluster name from the URL. If no cluster is defined in the URL, undefined is returned.
+ */
+export function useClusterFromURLVar(): string | undefined {
```
Comment (sniok) [inline frontend/src/lib/k8s/index.ts:148]: I think we should be very explicit about what this means, because we already have a cluster 'variable' in the url /c/cluster/. To disambiguate I think this should be named something like `useClusterFromQueryParam`

Comment (illume) [reply frontend/src/lib/k8s/index.ts:148]: What do you mean, never mind?

Comment (illume) [reply frontend/src/lib/k8s/index.ts:148]: I was thinking “query” is maybe overloaded with react query stuff.  But maybe it’s ok.

useClusterFromSearchParam is also not quite right making it sound like it’s from a search.

useClusterFromURLQuery?
useURLQueryCluster?

Comment (sniok) [reply frontend/src/lib/k8s/index.ts:148]: well there's a separate discussion about the url format, so this comment may not be relevant

Comment (sniok) [reply frontend/src/lib/k8s/index.ts:148]: useURLQueryCluster sounds good
hm or maybe an approach of naming it for what it is instead of where it's stored, something like `useDetailsSingleCluster`?

Comment (illume) [reply frontend/src/lib/k8s/index.ts:148]: Yeah maybe useDetailsSingleCluster is the least misdirecting/confusing.

Comment (illume) [reply frontend/src/lib/k8s/index.ts:148]: > well there's a separate discussion about the url format, so this comment may not be relevant, let me rephrase it better

Can you please let us know which proposal you’d prefer and why?

**Comment by sniok** [comment]
URL is supposed to be a unique resource identifier
if we think in terms of purely displaying details of a resource then `/c/SINGLE_CLUSTER/resource-kind/my-resource` makes sense
but the app also has selected clusters in the sidebar for example, so url has to represent that too

with the previous multi cluster changes we already changed the meaning of the /c/cluster/ param, it now represents all the selected clusters and not the cluster of a given resource. 

so now we need to store information about a given resource cluster:

we could repurpose the `/c/cluster/` param for it, but now it means different thing depending where we are in the app (selected clusters or cluster of a resource when showing details) and IMHO it's a bit confusing

we could put it in the query params `/c/clusetA+clusterB/pods/my-pod?cluster=clusterA` but query parameters are not supposed to be used for unique identification of a resource

a logical solution would be to introduce a new param in the url, like
`/clusterB+clusterA/details/clusterA/pods/my-pod` where `/{selectedclusters}/details/{resourceCluster}/{resourceKind}/{resourceName}`
(details can be ommited, variable order can be changed, it doesn't matter)

but this kind of solution seems like a big breaking change. so I think doing `?cluster=clusterA` is the simplest solution here?

File: frontend/src/lib/k8s/index.ts (hunk 1 above)
Comment (sniok) [reply frontend/src/lib/k8s/index.ts:148]: posted it in a separate comment

**Comment by illume** [comment]
> a logical solution would be to introduce a new param in the url, like
`/clusterB+clusterA/details/clusterA/pods/my-pod` where `/{selectedclusters}/details/{resourceCluster}/{resourceKind}/{resourceName}`
(details can be ommited, variable order can be changed, it doesn't matter)
> 
>  but this kind of solution seems like a big breaking change. so I think doing ?cluster=clusterA is the simplest solution here?

Adding `{resourceCluster}` into the URL could work too... I feel like it might even be the more correct choice?

How is it a breaking change though?

**Comment by illume** [comment]
> Detail links from current version wouldn't work in new version (if that is something we're concerned about)

I guess we can keep the previous route? So it will keep working as well.
- new `/{selectedclusters}/{resourceCluster}/details/{resourceKind}/{resourceName}`
- prev `/{selectedclusters}/details/{resourceKind}/{resourceName}`

One other con against ?cluster is that if you remove it.
- if you have `/{selectedclusters}/details/{resourceKind}/{resourceName}?cluster={resourceCluster}`
- then remove `?cluster?` the URL is broken: `/{selectedclusters}/details/{resourceKind}/{resourceName}`

So yeah... for these reasons I think it's worth trying out:
- new `/{selectedclusters}/{resourceCluster}/details/{resourceKind}/{resourceName}`
- prev `/{selectedclusters}/details/{resourceKind}/{resourceName}`

What do you think?

**Comment by illume** [comment]
@joaquimrocha can you please respond to this comment? https://github.com/headlamp-k8s/headlamp/pull/2520#issuecomment-2464610293

**Comment by illume** [comment]
Closing in favour of 
- https://github.com/kubernetes-sigs/headlamp/pull/3120
